### PR TITLE
Exclude shared block cache from core dump

### PR DESCRIPTION
**Pull request by wujy-cs**
## What have you changed? (mandatory)

Support jemalloc nodump allocator for shared block cache, so we can exclude block cache from core dump, which is useful to reduce core file size.

Add a memory allocator option to block cache options in case to support more allocator in future.

## What are the type of the changes? (mandatory)

New feature

## How has this PR been tested? (mandatory)

Deploy a one node tikv-server with ansible, load 100GB data, Then scan 100GB data with following scenario:

- 4GB block cache with jemalloc nodump allocator
- 4GB block cache with default allocator
- 10GB block cache with jemalloc nodump allocator
- 10GB block cache with default allocator

Dump tikv-server process after each scenario. The core file generated is 5.6GB, 10.6GB, 5.4GB,  18GB respectively.

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

Yes. We need to update config template.

## Refer to a related PR or issue link (optional)

#5161

### CODE REVIEW

**Comment by siddontang** [comment]
>  5.4GB, 18GB respectively.

the block cache is only 10 GB, but 18 - (5.4 + 10) = 2.6, seem it uses more 2.6 GB, right?

File: etc/config-template.toml [hunk 1]
```diff
@@ -158,6 +158,15 @@
...
 # capacity = "1GB"
 
+## Optional memory allocator to allocate shared block cache.
+## If set jemalloc nodump allocator, block cache will be excluded from core dump, which is useful to
+## reduce core file size. This allocator is available only if "jemalloc" feature is enabled.
+##
+## Default: not set, use default allocator of RocksDB
+## jemalloc nodump allocator: "jemalloc-nodump"
+##
+#  memory-allocator: "jemalloc-nodump"
```
Comment (siddontang) [inline etc/config-template.toml]: I think we can use it as default now, we don't need to add a configuration in the template here.

**Comment by siddontang** [comment]
how about the performance?

**Comment by wujy-cs** [comment]
> how about the performance?

Seems no difference. Here are some benchmarks:

Read 10000000 records, and Scan 10000 records for 10000 times on 100GB TiKV with go-ycsb, for following cases (change config by ansible rolling update):

Case 1.1:  4GB block cache, jemalloc nodump allocator
Case 1.2:  4GB block cache, default allocator
Case 2.1: 10GB block cache, jemalloc nodump allocator
Case 2.2: 10GB block cache, default allocator

Result :

Case 1.1:
- READ - Takes(s): 151.2, Count: 10000000, OPS: 66138.2, Avg(us): 478, Min(us): 93, Max(us): 62401, 95th(us): 1000.0, 99th(us): 2000.0      (Block cache hit 39%)
- SCAN - Takes(s): 121.1, Count: 9984, OPS: 82.4, Avg(us): 385163, Min(us): 66785, Max(us): 1880469, 95th(us): 493000.0, 99th(us): 522000.0     (Block cache hit 7.4%)

Case1.2:
- READ - Takes(s): 149.0, Count: 10000000, OPS: 67124.7, Avg(us): 471, Min(us): 97, Max(us): 58412, 95th(us): 1000.0, 99th(us): 2000.0      (Block cache hit 40.4%)
- SCAN - Takes(s): 119.3, Count: 9984, OPS: 83.7, Avg(us): 380242, Min(us): 75429, Max(us): 2219205, 95th(us): 483000.0, 99th(us): 510000.0     (Block cache hit 6.9%)

Case2.1:
READ - Takes(s): 147.2, Count: 10000000, OPS: 67937.7, Avg(us): 466, Min(us): 92, Max(us): 57841, 95th(us): 1000.0, 99th(us): 2000.0  (Block cache hit 47.8%)
SCAN - Takes(s): 121.1, Count: 9984, OPS: 82.5, Avg(us): 383553, Min(us): 68777, Max(us): 2370117, 95th(us): 500000.0, 99th(us): 551000.0   (Block cache hit 7.2%)

Case2.2:
- READ - Takes(s): 147.1, Count: 9999998, OPS: 67972.3, Avg(us): 465, Min(us): 98, Max(us): 277047, 95th(us): 1000.0, 99th(us): 2000.0   (Block cache hit 47.6%)
- SCAN - Takes(s): 121.6, Count: 9984, OPS: 82.1, Avg(us): 387245, Min(us): 68519, Max(us): 1619917, 95th(us): 499000.0, 99th(us): 538000.0   (Block cache hit 6.7%)

**Comment by wujy-cs** [comment]
> > 5.4GB, 18GB respectively.
> 
> the block cache is only 10 GB, but 18 - (5.4 + 10) = 2.6, seem it uses more 2.6 GB, right?

looks like so

File: etc/config-template.toml (hunk 1 above)
Comment (wujy-cs) [reply etc/config-template.toml]: Okay, just thought maybe we will support more allocator in the future.

**Comment by siddontang** [comment]
seem the default allocator is a little faster than nodump allocator.

**Comment by siddontang** [comment]
> looks like so

@wujy-cs , can we investigate it?

**Comment by wujy-cs** [comment]
> seem the default allocator is a little faster than nodump allocator.
@siddontang Previous results are not accurate since I did not clear page cache between each experiment. Here are some more experiments. For most result jemalloc-nodump allocator is faster than default allocator about 0.5%~1%. 
Default allocator occupied more virtual space after evicting expired cache, but physical memory usage is not changed.
https://docs.google.com/spreadsheets/d/1GakS6YnMIqjt9XbPiSxxyd6S7AuOHjIAZz75dTVYcWc/edit?usp=sharing

**Comment by wujy-cs** [comment]
> seem the default allocator is a little faster than nodump allocator.

@siddontang Previous results are not accurate since I did not clear page cache between each experiment. Here are some more experiments. For most result jemalloc-nodump allocator is faster than default allocator about 0.5%~1%.
Default allocator occupied more virtual space after evicting expired cache, but physical memory usage is not changed.
https://docs.google.com/spreadsheets/d/1GakS6YnMIqjt9XbPiSxxyd6S7AuOHjIAZz75dTVYcWc/edit?usp=sharing

File: etc/config-template.toml (hunk 1 above)
Comment (yiwu-arbug) [reply etc/config-template.toml]: Let's just have a `dump-block-cache: true|false` option, which is easier to understand.

**Comment by yiwu-arbug** [comment]
@siddontang do we need this for 3.0 branch? If so, maybe default to previous behavior (dump block cache) for 3.0?

**Comment by breeswish** [comment]
Can we also cherry pick to 2.1?

**Pull request merged**